Versioning.md

```diff
@@ -0,0 +1,83 @@
+# Versioning
+
+This document describes the versioning policy for this repository.
+
+## Goals
+
+### API Stability
+
+Once the API for a given signal (spans, logs, metrics, baggage) has been officially released, that API module will function with any SDK that has the same major version, and equal or greater minor or patch version.
+
+For example, libraries that are instrumented with `opentelemetry 1.0.1` will
+function in applications using `opentelemetry 1.11.33` or `opentelemetry
+1.3.4`.
+
+### SDK Stability
+
+Public portions of the SDK (constructors, configuration, end-user interfaces)
```

Comment:
I think we will run into problems when we add new methods or members to C++ class. It is inevitable that we'd end up doing some refactor for **v1.x**. Very bold, impractical and unrealistic expectation that we may freeze it in time and guarantee C++ API/ABI stability. While it may be easy to implement for other languages, we would have to be really creative and create some sort of enforcement via [ABI compliance checker](https://lvc.github.io/abi-compliance-checker/) , OR allow for some additional flexibility. Such as, allowing `ot::v150::TracerProvider` , `ot::v151::TracerProvider` in cases where we don't want to bump-up the MAJOR version. I read in the OTEP that it will take time to get to v2.. But that'll significantly constrain our abilities to address our customer feature needs --- given when we add new things, the newer version of the class may be breaking compatibility with old version. Once we release v1, we may need to have provision for inline namespaces such as v1_1, v1_2, etc. - where prebuilt SDK includes both, the original v1 class e.g. `v1::TracerProvider` , as well as `v1_2::TracerProvider` - to cover the cases where refactor may break the ABI.

Reply:
The API already has version namespacing built-in: https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/version.h#L11, so why not just have an OT_v1::ext<subversion>:: thing?

However, the notion of "extension" isn't clear here.   I agree with you we'd have to be creative if we find breaking changes in what we need to provide.    I personally think stability guarantees are more improtant than flexibility on our end, but it would be good to brainstorm and have a plan for how we can address changes in compatible ways.

Reply:
There are two aspects of ABI stability as I understand:
1. `sdk` compiled against one version of C++ standard library should work with application/library compiled against different compiler/or different version of compiler. We don't guarantee that for our SDK ( though we do it for API ). And as we are going to release source packages, the vendor's compiling and using this sdk need to ensure this compatibility if they want so.
2.  Doing `sdk` code changes which are not backward ABI compatible - this is something we are discussing here. Going through this document - https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C++  - It seems there are quite a few DON'T's to ensure the ABI stability. Would be good to go through them once and see if this is something doable once we reach first stable version **v1.0.** .  If it is too large an ask, we do need to resort to namespace level versioning which will bring lots of overhead and noise in code to maintain different version internally.